frontend/src/component/HeatMap.tsx

```diff
@@ -0,0 +1,85 @@
+import {ComposableMap, Geographies, Geography,} from "react-simple-maps";
+import {Country} from "../model/Country";
+import {useEffect, useState} from "react";
+import {scaleLinear} from "d3-scale";
+import useUsers from "../hook/useUsers";
+import {Traveller} from "../model/User";
+
+type HeatMapProps = {
+    countries: Country[]
+}
+const geoUrl = "https://raw.githubusercontent.com/deldersveld/topojson/master/world-countries.json"
+
+export default function HeatMap(props: HeatMapProps) {
+
+    const [maxDomain, setMaxDomain] = useState(1)
+
+    const colorScale = scaleLinear<string>()
+        .domain([0, maxDomain])
+        .range(["#f8bbd0", "#ad1457"]);
+
+    const countryList = props.countries
+    let userList:Traveller[] = useUsers().users
+
+    const [countryCounts, setCountryCounts] = useState<{
+        country: string,
+        count: number
+    }[]>([])
+
+    useEffect(() => {
+        const threeLetterCodeList = countryList.map(Country => Country.threeLetterCode)
+
+        function setCount(country: string) {
+            return {
+                country,
+                count: 0
+            };
+        }
+
+        let output = threeLetterCodeList.map(setCount);
+        for (const element of userList) {
+            for (const item of element.visitedCountries) {
+                const threeLetterCodeToCount = item.threeLetterCode
+                for (const element of output) if (element.country === threeLetterCodeToCount) {
+                    element.count++
+                }
+            }
+        }
+        let countVisits = []
+        for (const element of output) countVisits.push(element.count)
+        let maxVisits = Math.max(...countVisits)
+        setMaxDomain(maxVisits)
+        setCountryCounts(output)
+    }, [])
```

Comment:
Someone else then your probably needs a day to understand your code here. maybe less for-each and more filter(), map() etc?